### PR TITLE
Provide histograms for http server metrics

### LINKED ISSUE #28178: Provide histograms for http server metrics
### Description

To monitor a service level for Keycloak's API requests, Keycloak currently doesn't provide information about how many request are within or outside a service level range of for example a 100ms response time. 

### Motivation

An average response time is usually not how SLOs are defined. Instead, they as for 99% percent of the responses to be provided withing 100ms. 

This can be enabled in Quarkus as outlined in https://quarkus.io/guides/telemetry-micrometer with a MeterFilter.
Usually a user running Keycloak wants to monitor their response times to have some numbers available to monitor and also to show when customers complain.

In the Keycloak Benchmark Project, this allows us to then give a percentage of the responses within a SLO, but also allows us drawing heat maps of the response latency distribution over time. 

### Details

I'll prepare a PR.

### PR DESCRIPTION
**Pull request by ahus1**
Closes #28178

### CODE REVIEW

**Comment by ahus1** [comment]
@vmuzikar - I wonder if your team would be ok to add histograms for http responses, so that we'd be able to monitor SLOs for Keycloak's APIs in Keycloak Benchmark. Thanks!

File: quarkus/config-api/src/main/java/org/keycloak/config/HttpOptions.java [hunk 1]
```diff
@@ -132,4 +132,16 @@ public enum ClientAuth {
...
+    public static final Option<Boolean> HTTP_METRICS_HISTOGRAMS_ENABLED = new OptionBuilder<>("http-metrics-histograms-enabled", Boolean.class)
+            .category(OptionCategory.HTTP)
+            .description("Enables a histogram with default buckets for the duration of HTTP server requests.")
+            .defaultValue(Boolean.FALSE)
+            .build();
+
+    public static final Option<String> HTTP_METRICS_SLOS = new OptionBuilder<>("http-metrics-slos", String.class)
+            .category(OptionCategory.HTTP)
+            .description("Service level objectives for HTTP server requests. Use this instead of the default histogram, or use it in combination to add additional buckets. " +
+                    "Specify a list of comma-separated vales defined in milliseconds. Example with buckets from 5ms to 10s: 5,10,25,50,250,500,1000,2500,5000,10000")
```
Comment (mabartos) [inline quarkus/config-api/src/main/java/org/keycloak/config/HttpOptions.java:144]:
```suggestion
                    "Specify a list of comma-separated values defined in milliseconds. Example with buckets from 5ms to 10s: 5,10,25,50,250,500,1000,2500,5000,10000")
```

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/HttpPropertyMappers.java [hunk 2]
```diff
@@ -114,6 +114,13 @@ public static PropertyMapper<?>[] getHttpPropertyMappers() {
                 fromOption(HttpOptions.HTTP_POOL_MAX_THREADS)
                         .to("quarkus.thread-pool.max-threads")
                         .paramLabel("threads")
+                        .build(),
+                fromOption(HttpOptions.HTTP_METRICS_HISTOGRAMS_ENABLED)
```
Comment (mabartos) [inline quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/HttpPropertyMappers.java:118]: I was wondering about the semantics of these options as to whether we should gather all metrics-related stuff into an individual option category - Metrics. I understand it's related only to HTTP, but from the UX perspective, I'd choose to separate it as the metrics options might be hidden in chunks of other options. 

I can imagine a situation when user needs/wants to use and set some settings for metrics rather in a separate manner than setting of the HTTP server. To be more specific, I'd say the most common user case is to set the necessary stuff to have Keycloak operating and then add some additional things for manageability/observability - metrics.

I know we already have the `cache-metrics-histogram-enabled` option tied to the cache options and not sure what migration would need to be done (is it already part of some release?).

TL;DR I'd opt to keep metrics-related stuff in the `METRICS` category.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/services/metrics/HistogramMeterFilter.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+ * @author Alexander Schwartz
+ */
+@Singleton
+public class HistogramMeterFilter implements MeterFilter {
+
+    private boolean histogramsEnabled;
+    private double[] slos;
+
+    public HistogramMeterFilter() {
+        histogramsEnabled = Configuration.getOptionalBooleanKcValue(HttpOptions.HTTP_METRICS_HISTOGRAMS_ENABLED.getKey()).orElse(false);
```
Comment (mabartos) [inline quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/services/metrics/HistogramMeterFilter.java:41]:
```suggestion
        histogramsEnabled = Configuration.isTrue(HttpOptions.HTTP_METRICS_HISTOGRAMS_ENABLED);
```
jfyi - you can use `isTrue(Option<> o)`

**Review by mabartos — COMMENTED** [review]
@ahus1 Thanks for the PR! 

LGTM in general, I've just added one thing to consider, and some nitpicks.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/HttpPropertyMappers.java (hunk 2 above)
Comment (ahus1) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/HttpPropertyMappers.java:118]: I wouldn't mind either, and the cache metrics is scheduled on KC25 so it hasn't been released, yet. 

Looking at Quarkus, they took the approach to put `metrics-enabled` at the end, see `quarkus.scheduler.metrics.enabled` or `quarkus.hibernate-orm.metrics.enabled`

File: quarkus/config-api/src/main/java/org/keycloak/config/HttpOptions.java (hunk 1 above)
Comment (ahus1) [reply quarkus/config-api/src/main/java/org/keycloak/config/HttpOptions.java:144]: Thanks, fixed in next commit

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/HttpPropertyMappers.java (hunk 2 above)
Comment (mabartos) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/configuration/mappers/HttpPropertyMappers.java:118]: Considering it again, it'd probably make more sense to keep it as is. As you pointed to the Quarkus approach, it's quite natural for them to attach these properties to particular extensions - separated piece of work. And it make sense as we can think about these categories as the whole module as it should be delivered as one big piece. Basically, without HTTP "module", there would not be any HTTP metrics, etc.

So, +1 keep it as is.

**Review by mabartos — APPROVED** [review]
@ahus1 LGTM

**Review by hmlnarik — APPROVED** [review]
Approving per @mabartos 's review

**Comment by ahus1** [comment]
This is sharing some of the heatmaps created from histograms here which we capture for a cross-site setup [in the Keycloak Benchmark project here](https://github.com/keycloak/keycloak-benchmark/blob/a5b20a15164f22234be931f36776ac2b4f55bda8/provision/minikube/monitoring/dashboards/authentication-code.json): 

image

The left image shows where we met our SLO of 250 ms, the second show the distribution of response times as a heatmap, basically logarithmically scaled, and the third diagrams shows the 99% percentile over time for different URLs.

Setup is as usual two sites HA setup, with clustered Infinispan and an Aurora Multi-AZ database.
